src/main/java/com/iota/iri/hash/Kerl.java

```diff
@@ -6,142 +6,174 @@
             //copy trits[offset:offset+length]
-            System.arraycopy(trits, offset, trit_state, 0, HASH_LENGTH);
+            System.arraycopy(trits, pos, trit_state, 0, HASH_LENGTH);
 
             //convert to bits
             trit_state[HASH_LENGTH - 1] = 0;
-            bytesFromBigInt(bigIntFromTrits(trit_state, 0, HASH_LENGTH), byte_state, 0);
+            bytesFromBigInt(bigIntFromTrits(trit_state, 0, HASH_LENGTH), byte_state);
 
             //run keccak
             keccak.update(byte_state);
-            offset += HASH_LENGTH;
-
-        } while ((length -= HASH_LENGTH) > 0);
+        }
     }
 
     @Override
-    public void squeeze(final int[] trits, int offset, int length) {
-
+    public void squeeze(final int[] trits, final int offset, final int length) {
         if (length % 243 != 0) {
-            throw new RuntimeException("Illegal length: " + length);
+            throw new IllegalArgumentException("Illegal length: " + length);
         }
-
         try {
-          do {
-              this.keccak.digest(byte_state, 0, BYTE_HASH_LENGTH);
-              //convert to trits
-              tritsFromBigInt(bigIntFromBytes(byte_state, 0, BYTE_HASH_LENGTH), trit_state, 0, Sponge.HASH_LENGTH);
-
-              //copy with offset
-              trit_state[HASH_LENGTH - 1] = 0;
-              System.arraycopy(trit_state, 0, trits, offset, HASH_LENGTH);
-
-              //calculate hash again
-              for (int i = byte_state.length; i-- > 0; ) {
-
-                  byte_state[i] = (byte) (byte_state[i] ^ 0xFF);
-              }
-              keccak.update(byte_state);
-              offset += HASH_LENGTH;
-
-          } while ((length -= HASH_LENGTH) > 0);
+            for (int pos = offset; pos < offset + length; pos += HASH_LENGTH) {
+                this.keccak.digest(byte_state, 0, BYTE_HASH_LENGTH);
+                //convert to trits
+                tritsFromBigInt(bigIntFromBytes(byte_state, 0, BYTE_HASH_LENGTH), trit_state, 0, Sponge.HASH_LENGTH);
+
+                //copy with offset
+                trit_state[HASH_LENGTH - 1] = 0;
+                System.arraycopy(trit_state, 0, trits, pos, HASH_LENGTH);
+
+                //calculate hash again
+                for (int i = byte_state.length; i-- > 0; ) {
+                    byte_state[i] = (byte) (byte_state[i] ^ 0xFF);
+                }
+                keccak.update(byte_state);
+            }
         } catch (DigestException e) {
-            e.printStackTrace();
+            e.printStackTrace(System.err);
             throw new RuntimeException(e);
         }
     }
 
     public static BigInteger bigIntFromTrits(final int[] trits, final int offset, final int size) {
-
+        for (int i = offset; i < offset + size; i++) {
+            if (trits[i] < -1 || trits[i] > 1) {
+                throw new IllegalArgumentException("not a trit: " + trits[i]);
+            }
+        }
         BigInteger value = BigInteger.ZERO;
```

Comment:
This check really isn't required from trusted data. There isn't checking anywhere else ... but I thought it was good form to include it. Ideally, all the trit[] should be checked on first input. Tests show that the check doesn't really slow anything down.